models.js

```diff
@@ -482,103 +555,147 @@ debuglog = function () {
         this.toString = binOpToString;
         this.toXPath = binOpToXPath;
+        this.toHashtag = binOpToHashtag.bind(this);
         this.getChildren = binOpChildren;
         return this;
     };
-    
+
     xpm.XPathCmpExpr = function(definition) {
         this.type = definition.type;
         this.left = definition.left;
         this.right = definition.right;
         this.toString = binOpToString;
         this.toXPath = binOpToXPath;
+        this.toHashtag = binOpToHashtag.bind(this);
         this.getChildren = binOpChildren;
         return this;
     };
-    
+
     xpm.XPathArithExpr = function(definition) {
         this.type = definition.type;
         this.left = definition.left;
         this.right = definition.right;
         this.toString = binOpToString;
         this.toXPath = binOpToXPath;
+        this.toHashtag = binOpToHashtag.bind(this);
         this.getChildren = binOpChildren;
         return this;
     };
-    
+
     xpm.XPathUnionExpr = function(definition) {
         this.type = definition.type;
         this.left = definition.left;
         this.right = definition.right;
         this.toString = binOpToString;
         this.toXPath = binOpToXPath;
+        this.toHashtag = binOpToHashtag.bind(this);
         this.getChildren = binOpChildren;
         return this;
     };
-    
+
     xpm.XPathNumNegExpr = function(definition) {
         this.type = definition.type;
         this.value = definition.value;
         this.toString = function() {
             return "{unop-expr:" + this.type + "," + String(this.value) + "}";
         };
-        this.toXPath = function() {
-            return "-" + this.value.toXPath();
-        };
+        function _combine(transFunc) {
+            return function() {
+                return "-" + transFunc(this.value);
+            };
+        }
+        this.toXPath = _combine(objToXPath);
+        this.toHashtag = _combine(objToHashtag);
         this.getChildren = function () {
            return [this.value];
         };
         return this;
     };
-}());
+
+    return xpm;
+};
+
+var defaultHashtagConfig = {
+    // @param namespace - the namespace used in hashtag
+    // @return - truthy value
+    isValidNamespace: function (namespace) {
+        return false;
+    },
+    // @param hashtagExpr - text of hashtag ex. #form/question
+    // @return - the XPath or falsy value if no corresponding XPath found
+    hashtagToXPath: function (hashtagExpr) {
+        throw new Error("This should be overridden");
+    },
+    // @param xpath_ - XPath object (can be any of the objects defined in xpm
+    // @returns - text representation of XPath in hashtag format (default
+    //            implementation is to just return the XPath)
+    toHashtag: function (xpath_) {
+        return xpath_.toXPath();
+    },
+};
+
+var xpathmodels = XPathModels(defaultHashtagConfig);
+xpathmodels.DEBUG_MODE = false;
+
+debuglog = function () {
+    if (xpathmodels.DEBUG_MODE) {
```

Comment:
Looks like this only works for the default `xpathmodels` instance. Would it make sense to move this inside `XPathModels`? That way setting `inst.DEBUG_MODE = true` on any `XPathModels` instance would do what you expect.